compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5902,7 +5902,8 @@ private BType getQueryMapConstraintType(BType type, Location pos) {
             }
         } else if (type.tag == TypeTags.TUPLE) {
             List<BType> tupleTypeList = ((BTupleType) type).tupleTypes;
-            if (tupleTypeList.size() == 2 && TypeTags.isStringTypeTag(tupleTypeList.get(0).tag)) {
+            if (tupleTypeList.size() == 2 &&
+                    TypeTags.isStringTypeTag(Types.getReferredType(tupleTypeList.get(0)).tag)) {
```

Comment:
Still won't work for subtypes of string, right?

```ballerina
type FooBar "foo"|"bar";

function test() {
    [FooBar, int|float][] list = [["foo", 1.4], ["bar", 2], ["foo", 3]];
    map<int|float> & readonly|error mp1 = map from var item in list select item; // incompatible type in select clause: expected [string,any|error], found '[FooBar,(int|float)]'
}
```

Reply:
will check.

Reply:
Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/37402/commits/fa3d9b61f3e78f304e6ae2495183e2d58e77af98.